### PR TITLE
fix: improve terminal marker placements on windows

### PR DESCRIPTION
**Pull request by cpendery**
This PR attempts to approve terminal marker placements on Windows in 2 ways.

1) If the Windows prompt adjustment fails, reattempt to identify the command start when a user provides input. This helps with multiline prompts like used in Git Bash
2) Adjust the prompt start and end markers for all Windows prompts, not just single line prompts
  *note:* while this will result in the starting line of a multiline prompt getting captured in the output, it improves the sticky scroll behavior significantly.

### Sticky Scroll Pre Fix
https://github.com/microsoft/vscode/assets/35637443/ece832ff-abad-4110-9f24-39058e2841d0

### Sticky Scroll Post Fix
https://github.com/microsoft/vscode/assets/35637443/e68cea18-a2c3-473e-b3b0-094acf36b701

### Input Prompt Adjustment Example 1
https://github.com/microsoft/vscode/assets/35637443/0b5abbfc-ff47-4cdc-8c27-3f25cad2e447

### Input Prompt Adjustment Example 2
https://github.com/microsoft/vscode/assets/35637443/b4547168-3f8f-4dab-9fc5-7471730dec60

Related to #143769 & #209003

### CODE REVIEW

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts [hunk 1]
```diff
@@ -661,21 +681,24 @@ class WindowsPtyHeuristics extends Disposable {
...
-						this._capability.currentCommand.promptStartMarker?.dispose();
-						this._capability.currentCommand.promptStartMarker = cloneMarker(this._terminal, this._capability.currentCommand.commandStartMarker);
-						// Adjust the last command if it's not in the same position as the following
-						// prompt start marker
-						const lastCommand = this._capability.commands.at(-1);
-						if (lastCommand && this._capability.currentCommand.commandStartMarker.line !== lastCommand.endMarker?.line) {
-							lastCommand.endMarker?.dispose();
-							lastCommand.endMarker = cloneMarker(this._terminal, this._capability.currentCommand.commandStartMarker);
-						}
+
```
Comment (Tyriar) [inline src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:684]: This is a no go imo, to give you an idea of what I see here:

Before:

Recording 2024-04-01 at 04 26 04

After:

Recording 2024-04-01 at 04 26 44

It's a little hard to see in the gif but it also drives other smaller things like the grey bar on hover:

image

The blue bar on command nav (ctrl+up):

image

(very recent change) Also in run recent command (ctrl+alt+r):

image

---

Is there any way we can get the best of both worlds here? For example we could customize the `_tryAdjustCommandStartMarker` behavior when it comes via `handleInput`?

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts [hunk 2]
```diff
@@ -600,6 +607,19 @@ class WindowsPtyHeuristics extends Disposable {
 		}
 	}
 
+	handleInput() {
```
Comment (Tyriar) [inline src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:610]: Good idea to adjust on input, we need to keep in mind that `onData` could happen when not in a prompt though. We should probably only do this when prompt start and command start has happened, but not command executed

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts (hunk 1 above)
Comment (cpendery) [reply src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:684]: @Tyriar The before and after gifs look very similar to me at this speed, what is the major difference between the two? Happy to adjust the `_tryAdjustCommandStartMarker` for each case

Comment (Tyriar) [reply src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:684]: The major difference is that the powerline part is included in the sticky scroll too, but importantly not the initial empty line of the prompt.

Comment (cpendery) [reply src/vs/platform/terminal/common/capabilities/commandDetectionCapability.ts:728]: I adjusted the logic to actually pull the prompt height during the prompt, similar to the `cwd` to allow for proper prompt adjustment and it seems to be working much better for the multiline prompts. Do you think this is an acceptable approach? I'm occasionally getting left over markers with single line powershell that I can work on if you like it

## Working Examples
https://github.com/microsoft/vscode/assets/35637443/2190db89-5010-42ea-af22-eed77729c570


https://github.com/microsoft/vscode/assets/35637443/5d85ab16-fcf9-4ec9-baa2-6b6586a69d0b

File: src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh [hunk 3]
```diff
@@ -171,6 +171,11 @@ __vsc_update_cwd() {
 	builtin printf '\e]633;P;Cwd=%s\a' "$(__vsc_escape_value "$PWD")"
 }
 
+__vsc_update_prompt_height() {
+	__vsc_prompt_height="$(builtin printf "%s" "${PS1@P}" | wc -l)"
+	builtin printf '\e]633;P;PromptHeight=%s\a' "$(__vsc_escape_value "$__vsc_prompt_height")"
+}
```
Comment (Tyriar) [inline src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:177]: Neat idea, we probably have to make this a lower-bound for the prompt height though due to wrapping. See here how the VS Code thinks it's single line (bash/macOS):

Screenshot 2024-04-16 at 4 10 31 PM

**Review by Tyriar — CHANGES_REQUESTED** [review]

File: src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh (hunk 3 above)
Comment (cpendery) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:177]: Could we set a condition that we only do this additional input adjustment if all the lines the prompt range (current + promptHeight) are not wrapped using xterm's `isWrapped` property?

Comment (Tyriar) [reply src/vs/workbench/contrib/terminal/browser/media/shellIntegration-bash.sh:177]: That will probably work 👍

**Comment by cpendery** [comment]
Alright, I fixed the stray markers, avoid adjusting the prompt start if the text is wrapped, and 1-2 small issues I found testing. Here are some new demo videos

### No Adjust When Wrapped
https://github.com/microsoft/vscode/assets/35637443/5eabbb42-43f7-43c3-86b4-b0406dbceb12
### Multi-line Git Bash Adjust
https://github.com/microsoft/vscode/assets/35637443/1789d1da-a942-4f9d-b6f5-46fcb4202b3c
### Multi-line Pwsh Adjust
https://github.com/microsoft/vscode/assets/35637443/37c42a48-541f-4e94-a8ff-ef4baa0d8943
### Single-line Pwsh Adjust
https://github.com/microsoft/vscode/assets/35637443/1cf0532c-9a98-4ab7-9256-25f8f21f144c

**Comment by Tyriar** [comment]
Just tried with Windows/pwsh 7/starship and the multi-line prompt is still not coming through in sticky scroll:

image

**Comment by cpendery** [comment]
> Just tried with Windows/pwsh 7/starship and the multi-line prompt is still not coming through in sticky scroll:
> 
> image
> 
> It's working for git bash though 🤔

I'll setup starship and see if I can replicate this.

**Comment by cpendery** [comment]
I think I fixed it, there were small bugs. It's working for git bash (starship, default multiline, singleline) &  pwsh (starship, oh-my-posh [json theme], & default)

1) bash prompts were getting a height of 1 too small since it doesn't have trailing newlines like a powershell prompt
2) I used an offset of the prompt height, but it should be promptHeight -1 since the current line is part of the prompt
3) The bounds were wrong for cloning a marker below line 0 of the terminal, the offset should be `- cursorY` instead of `0`

**Comment by Tyriar** [comment]
Still the same as above

main

image

pr branch:

image

Will investigate on my end

**Comment by Tyriar** [comment]
After running an empty command then ls 👀 

image

**Comment by Tyriar** [comment]
Oh I was testing an old version 🤦

**Review by Tyriar — DISMISSED** [review]
Looking good!

**Review by Tyriar — APPROVED** [review]

**Review by lramos15 — APPROVED** [review]

**Pull request merged**